torrt/trackers/ytsmx.py

```diff
@@ -0,0 +1,146 @@
+from typing import List, Dict, Tuple, Optional
+
+from bs4 import BeautifulSoup
+from bs4.element import Tag
+from requests.models import Response
+
+from ..exceptions import TorrtTrackerException
+from ..base_tracker import GenericPublicTracker
+
+
+API_BASE = "https://yts.mx/api/v2/"
+QualityLinks = Dict[str, str]
+QualityLink = Tuple[str, str]
+
+class YtsmxTrackerException(TorrtTrackerException):
+    """Yts.mx specific exceptions"""
+
+class YtsmxTracker(GenericPublicTracker):
+    """This class implements .torrent files downloads for https://yts.mx tracker."""
+
+    alias: str = 'yts.mx'
+
+    test_urls: List[str] = [
+        'https://yts.mx/movies/the-matrix-resurrections-2021',
+    ]
+
+    def __init__(self, quality_prefs: List[str] = None):
+        super().__init__()
+
+        if quality_prefs is None:
+            quality_prefs = ['1080P.WEB', '720P.WEB']
+        else:
+            quality_prefs = [self._sanitize_quality(x) for x in quality_prefs]
+
+        self.quality_prefs = quality_prefs
+
+    @staticmethod
+    def _sanitize_quality(q: str) -> str:
+        """sanitizes quality provided by configuration"""
+
+        return q.strip().upper()
+
+    @staticmethod
+    def _get_quality_from_torrent(torrent: dict) -> str:
+        """returns sanitized quality string from API's torrent dict"""
+
+        return f"{torrent['quality']}.{torrent['type']}".upper()
+
+    def _get_torrent_page(self, url: str) -> BeautifulSoup:
+        """loads and parses main torrent's page"""
+
+        soup = self.get_torrent_page(url)
+        if not soup:
+            raise YtsmxTrackerException("main torrent page loading failed")
+
+        return soup
+
+    def _extract_movie_id(self, root: BeautifulSoup) -> str:
+        """returns movie id from tracker's info page"""
+
+        movie_info_tag = root.find("div", {"id": "movie-info"})
+        if not movie_info_tag:
+            raise YtsmxTrackerException("div#movie-info not found on page")
+
+        movie_info_tag: Tag
+        movie_id = movie_info_tag.attrs['data-movie-id']
+        if not movie_id.isdigit():
+            raise YtsmxTrackerException("movie-id is not a digit. Markup is changed")
+
+        return movie_id
+
+    def _get_movie_details(self, movie_id: str) -> dict:
+        """returns dict with movie info by calling API"""
+
+        response = self.get_response(API_BASE + f"movie_details.json?movie_id={movie_id}")
+        if not response:
+            raise YtsmxTrackerException("API didn't respond")
+
+        response: Response
```

Comment:
Is it necessary? You may want to  type hint base `.get_response()` instead.

Reply:
Yea, in this case, I think it's necessary, as during `.get_response` it's still of type of `Optional[Response]` (eg. `None | Response`).
So I think this line clarifies the type we are working with below. Explicitly.